x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/apikey/RestGetMyApiKeyAction.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.rest.action.apikey;
+
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.license.XPackLicenseState;
+import org.elasticsearch.rest.BytesRestResponse;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.rest.action.RestBuilderListener;
+import org.elasticsearch.xpack.core.security.action.GetApiKeyResponse;
+import org.elasticsearch.xpack.core.security.action.GetMyApiKeyAction;
+import org.elasticsearch.xpack.core.security.action.GetMyApiKeyRequest;
+import org.elasticsearch.xpack.security.rest.action.SecurityBaseRestHandler;
+
+import java.io.IOException;
+
+/**
+ * Rest action to get information for one or more API keys owned by the authenticated user.
+ */
+public class RestGetMyApiKeyAction extends SecurityBaseRestHandler {
+
+    public RestGetMyApiKeyAction(Settings settings, RestController controller, XPackLicenseState licenseState) {
+        super(settings, licenseState);
+        controller.registerHandler(RestRequest.Method.GET, "/_security/api_key/my", this);
```

Comment:
I understand your reasoning and you're right, but I still think we can use the authz framework we have and still have a single REST endpoint. Let me try to detail it here:

*  Looking at `RestGetApiKeyAction#innerPrepareRequest`, it is building the`GetApiKeyRequest`, and then calls the transport action that will be authorized. The `username` parameter is optional (as are all the others). I propose we "infer" the `username` parameter, _in case it is not specified in the request_ (otherwise we leave it be).
* The `username` will be `null` (or `*` or `_all` depending on the interface we decide further) if the user has the `manage_api_key` privilege or `Authentication.getAuthentication(threadPool.getThreadContext()).getUser().principal()` otherwise. To check the privilege we make a _has privilege_ call.
* We would then tweak the `RBACEngine` to do **the same thing** it does for `UserRequests` but for `GetApiKey` and `InvalidateApiKey`. Basically, authorize the action if the request has the same name as the principal.
* at the `ApiKeyService` level, call `findApiKeysForUserRealmApiKeyIdAndNameCombination` . Note that it is no longer correct to check the
```
        if (Strings.hasText(apiKeyId) || Strings.hasText(apiKeyName)) {
            if (Strings.hasText(realmName) || Strings.hasText(userName)) {
                listener.onFailure(new IllegalArgumentException(
                        "username or realm name must not be specified when the api key id or api key name is specified"));
            }
        }
```
condition at this level, because the `username` can be specified by the Rest handler, as detailed. This validation should be performed in the REST handler.

What do you think?

Reply:
I haven't had time to digest @albertzaharovits's proposal, and I want to submit my review comments, so let me just say:
- I would _love_ to not have too many new Rest APIs for this. None is great, but even 1 new API is better 2 or 3.
- _But_ we do need:
    - a security model that makes sense and we can review & understand
    - a user-level API that is clear enough that we can document it (including the security rules)

So if we need a new REST API to satisfy those last 2 requirements, then so be it. But if we can avoid it, that's great.